plugins/auth-backend-module-oidc-provider/src/authenticator.ts

```diff
@@ -0,0 +1,181 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Issuer,
+  ClientAuthMethod,
+  TokenSet,
+  UserinfoResponse,
+  Strategy as OidcStrategy,
+} from 'openid-client';
+import {
+  createOAuthAuthenticator,
+  PassportOAuthAuthenticatorHelper,
+  PassportOAuthDoneCallback,
+} from '@backstage/plugin-auth-node';
+
+/** @public */
+export const oidcAuthenticator = createOAuthAuthenticator({
+  defaultProfileTransform:
+    PassportOAuthAuthenticatorHelper.defaultProfileTransform,
+  async initialize({ callbackUrl, config }) {
+    const clientId = config.getString('clientId');
+    const clientSecret = config.getString('clientSecret');
+    const metadataUrl = config.getString('metadataUrl');
+    const customCallbackUrl = config.getOptionalString('callbackUrl');
+    const callbackUrl2 = customCallbackUrl || callbackUrl;
+    const tokenEndpointAuthMethod = config.getOptionalString(
+      'tokenEndpointAuthMethod',
+    ) as ClientAuthMethod;
+    const tokenSignedResponseAlg = config.getOptionalString(
+      'tokenSignedResponseAlg',
+    );
+    const initializedScope = config.getOptionalString('scope');
+    const initializedPrompt = config.getOptionalString('prompt');
+
+    const issuer = await Issuer.discover(metadataUrl);
+    const client = new issuer.Client({
+      access_type: 'offline', // this option must be passed to provider to receive a refresh token
+      client_id: clientId,
+      client_secret: clientSecret,
+      redirect_uris: [callbackUrl2],
+      response_types: ['code'],
+      token_endpoint_auth_method:
+        tokenEndpointAuthMethod || 'client_secret_basic',
+      id_token_signed_response_alg: tokenSignedResponseAlg || 'RS256',
+      scope: initializedScope || '',
+    });
+
+    const strategy = new OidcStrategy(
+      {
+        client,
+        passReqToCallback: false,
+      },
+      (
+        tokenset: TokenSet,
+        userinfo: UserinfoResponse,
+        done: PassportOAuthDoneCallback,
+      ) => {
+        if (typeof done !== 'function') {
+          throw new Error(
+            'OIDC IdP must provide a userinfo_endpoint in the metadata response',
+          );
+        }
+
+        const emails = userinfo.email ? [{ value: userinfo.email }] : undefined;
+        const photos = userinfo.picture
+          ? [{ value: userinfo.picture }]
+          : undefined;
+        const name =
+          userinfo.family_name && userinfo.given_name
+            ? {
+                familyName: userinfo.family_name,
+                givenName: userinfo.given_name,
+                middleName: userinfo.middle_name,
+              }
```

Comment:
Not blocking: not sure exactly where this comes from, but it gives me pause to see that we only check for given + family but then return given + family + middle? I think this was copy-pasted from the old `makeProfileInfo`...

Reply:
Hmm what do you suggest we do for this? if i remember correctly this was me looking at the UserInfo type and seeing that there were a ton of optional values that could populate the object, so i was trying to include as many of those values as possible.